File: go/subject.go
package beval

// NormalizeSubject creates a normalized Subject from system output. See SPEC §7.2.
//
// The returned Subject provides a consistent interface for graders regardless
// of the underlying system implementation.
func NormalizeSubject(
	query string,
	answer string,
	completionTime float64,
	opts ...SubjectOption,
) Subject {
	s := Subject{
		Query:          query,
		Answer:         answer,
		CompletionTime: completionTime,
		SourcesUsed:    []string{},
		Spans:          []interface{}{},
		Metadata:       map[string]interface{}{},
	}
	for _, opt := range opts {
		opt(&s)
	}
	return s
}

// SubjectOption configures optional fields on a Subject.
type SubjectOption func(*Subject)

// WithDocumentsRetrieved sets the documents retrieved count.
func WithDocumentsRetrieved(n int) SubjectOption {
	return func(s *Subject) {
		s.DocumentsRetrieved = n
	}
}

// WithCitationsCount sets the citations count.
func WithCitationsCount(n int) SubjectOption {
	return func(s *Subject) {
		s.CitationsCount = n
	}
}

// WithSourcesUsed sets the sources used list.
func WithSourcesUsed(sources []string) SubjectOption {
	return func(s *Subject) {
		s.SourcesUsed = sources
	}
}

// WithMetadata sets additional metadata on the subject.
func WithMetadata(meta map[string]interface{}) SubjectOption {
	return func(s *Subject) {
		s.Metadata = meta
	}
}

File: typescript/src/subject.ts
/**
 * Subject normalization for the beval framework.
 *
 * Wraps system output into the normalized Subject interface.
 * See SPEC.md §7.2 (The Subject).
 */

import type { Subject } from "./types.js";

/** Options for creating a normalized Subject. */
export interface NormalizeSubjectOptions {
  query: string;
  answer: string;
  completionTime: number;
  documentsRetrieved?: number;
  citationsCount?: number;
  sourcesUsed?: string[];
  spans?: unknown[];
  metadata?: Record<string, unknown>;
}

/**
 * Create a normalized Subject from system output. See SPEC §7.2.
 *
 * The returned Subject is a frozen object to prevent graders from
 * modifying shared state.
 */
export function normalizeSubject(options: NormalizeSubjectOptions): Readonly<Subject> {
  return Object.freeze({
    query: options.query,
    answer: options.answer,
    completionTime: options.completionTime,
    documentsRetrieved: options.documentsRetrieved ?? 0,
    citationsCount: options.citationsCount ?? 0,
    sourcesUsed: Object.freeze([...(options.sourcesUsed ?? [])]),
    spans: Object.freeze([...(options.spans ?? [])]),
    metadata: Object.freeze({ ...(options.metadata ?? {}) }),
  });
}
